File: OMSimulator/PI_Controller.lua
-- status: correct
-- teardown_command: rm -rf PI_Controller-lua.log PI_Controller-lua_tmp/ PI_Controller_*.dot PI_Controller_*.pdf
-- linux: yes
-- mingw: no

function exitOnError (status)
  print("status code: " .. status)
  if 0 ~= status then
    os.exit(0)
  end
end

oms2_setLoggingLevel(1)
status = 0;

oms2_setLogFile("PI_Controller-lua.log")
status = oms2_setTempDirectory("./PI_Controller-lua_tmp")
exitOnError(status)

status = oms2_newFMIModel("PI_Controller")
exitOnError(status)

-- instantiate FMUs
status = status + oms2_addFMU("PI_Controller", "../FMUs/Modelica.Blocks.Math.Add.fmu", "addP")
status = status + oms2_addFMU("PI_Controller", "../FMUs/Modelica.Blocks.Math.Gain.fmu", "P")
status = status + oms2_addFMU("PI_Controller", "../FMUs/Modelica.Blocks.Math.Add3.fmu", "addI")
status = status + oms2_addFMU("PI_Controller", "../FMUs/Modelica.Blocks.Continuous.Integrator.fmu", "I")
status = status + oms2_addFMU("PI_Controller", "../FMUs/Modelica.Blocks.Math.Add.fmu", "addPI")
status = status + oms2_addFMU("PI_Controller", "../FMUs/Modelica.Blocks.Math.Gain.fmu", "gainPI")
status = status + oms2_addFMU("PI_Controller", "../FMUs/Modelica.Blocks.Nonlinear.Limiter.fmu", "limiter")
status = status + oms2_addFMU("PI_Controller", "../FMUs/Modelica.Blocks.Math.Add.fmu", "addSat")
status = status + oms2_addFMU("PI_Controller", "../FMUs/Modelica.Blocks.Math.Gain.fmu", "gainTrack")
exitOnError(status)

-- add connections
status = status + oms2_addConnection("PI_Controller", "addP:y", "P:u")
status = status + oms2_addConnection("PI_Controller", "addI:y", "I:u")
status = status + oms2_addConnection("PI_Controller", "P:y", "addPI:u1")
status = status + oms2_addConnection("PI_Controller", "I:y", "addPI:u2")
status = status + oms2_addConnection("PI_Controller", "addPI:y", "gainPI:u")
status = status + oms2_addConnection("PI_Controller", "gainPI:y", "limiter:u")
status = status + oms2_addConnection("PI_Controller", "gainPI:y", "addSat:u2")
status = status + oms2_addConnection("PI_Controller", "limiter:y", "addSat:u1")
status = status + oms2_addConnection("PI_Controller", "addSat:y", "gainTrack:u")
status = status + oms2_addConnection("PI_Controller", "gainTrack:y", "addI:u3")
exitOnError(status)

-- define parameters
k = 100.0
yMax = 12.0
yMin = -yMax
wp = 1.0
Ni = 0.1
xi_start = 0.0

-- set parameters
status = status + oms2_setRealParameter("PI_Controller.addP:k1", wp)
status = status + oms2_setRealParameter("PI_Controller.addP:k2", -1.0)
status = status + oms2_setRealParameter("PI_Controller.addI:k2", -1.0)
status = status + oms2_setRealParameter("PI_Controller.I:y_start", xi_start)
status = status + oms2_setRealParameter("PI_Controller.gainPI:k", k)
status = status + oms2_setRealParameter("PI_Controller.limiter:uMax", yMax)
status = status + oms2_setRealParameter("PI_Controller.addSat:k2", -1.0)
status = status + oms2_setRealParameter("PI_Controller.gainTrack:k", 1.0/(k*Ni))
exitOnError(status)

-- simulation settings
status = status + oms2_setStartTime("PI_Controller", 0.0)
status = status + oms2_setStopTime("PI_Controller", 4.0)
status = status + oms2_setCommunicationInterval("PI_Controller", 1e-4)
status = status + oms2_setResultFile("PI_Controller", "PI_Controller.mat")
exitOnError(status)

status = status + oms2_exportCompositeStructure("PI_Controller", "PI_Controller_CompositeStructure.dot")
exitOnError(status)
os.execute("dot -Gsplines=none PI_Controller_CompositeStructure.dot | neato -n -Gsplines=ortho -Tpdf -oPI_Controller_CompositeStructure.pdf")

-- instantiate lookup table
status = status + oms2_addTable("PI_Controller", "setpoint.csv", "setpoint")
status = status + oms2_addTable("PI_Controller", "drivetrain.csv", "driveTrain")
exitOnError(status)

-- add connections
status = status + oms2_addConnection("PI_Controller", "setpoint:speed", "addP:u1")
status = status + oms2_addConnection("PI_Controller", "setpoint:speed", "addI:u1")
status = status + oms2_addConnection("PI_Controller", "driveTrain:w", "addP:u2")
status = status + oms2_addConnection("PI_Controller", "driveTrain:w", "addI:u2")
exitOnError(status)

status = status + oms2_exportCompositeStructure("PI_Controller", "PI_Controller_CompositeStructure.dot")
exitOnError(status)
os.execute("dot -Gsplines=none PI_Controller_CompositeStructure.dot | neato -n -Gsplines=ortho -Tpdf -oPI_Controller_CompositeStructure.pdf")

-- initialize
status = status + oms2_initialize("PI_Controller")
exitOnError(status)

status = status + oms2_exportDependencyGraphs("PI_Controller", "PI_Controller_initialUnknowns.dot", "PI_Controller_outputs.dot")
exitOnError(status)
os.execute("gvpr -c \"N[$.degree==0]{delete(root, $)}\" PI_Controller_initialUnknowns.dot | dot -Tpdf -o PI_Controller_initialUnknowns.pdf")
os.execute("gvpr -c \"N[$.degree==0]{delete(root, $)}\" PI_Controller_outputs.dot | dot -Tpdf -o PI_Controller_outputs.pdf")

-- simulate
status = status + oms2_simulate("PI_Controller")
status = status + oms2_terminate("PI_Controller")
exitOnError(status)

status = oms2_unloadModel("PI_Controller")
exitOnError(status)

-- Result:
-- status code: 0
-- status code: 0
-- status code: 0
-- status code: 0
-- status code: 0
-- status code: 0
-- status code: 0
-- status code: 0
-- status code: 0
-- status code: 0
-- status code: 0
-- status code: 0
-- status code: 0
-- status code: 0
-- info:    Logging information has been saved to "PI_Controller-lua.log"
-- endResult

File: OMSimulator/PI_Controller.py
## status: correct
## teardown_command: rm PI_Controller.log  PI_Controller.mat PI_Controller*.dot

from OMSimulator import OMSimulator
session = OMSimulator()

session.setLogFile("PI_Controller.log")
session.setTempDirectory("./PI_Controller_tmp")

model = session.newModel()

# instantiate FMUs
session.instantiateFMU(model, "../FMUs/Modelica.Blocks.Math.Add.fmu", "addP")
session.instantiateFMU(model, "../FMUs/Modelica.Blocks.Math.Gain.fmu", "P")
session.instantiateFMU(model, "../FMUs/Modelica.Blocks.Math.Add3.fmu", "addI")
session.instantiateFMU(model, "../FMUs/Modelica.Blocks.Continuous.Integrator.fmu", "I")
session.instantiateFMU(model, "../FMUs/Modelica.Blocks.Math.Add.fmu", "addPI")
session.instantiateFMU(model, "../FMUs/Modelica.Blocks.Math.Gain.fmu", "gainPI")
session.instantiateFMU(model, "../FMUs/Modelica.Blocks.Nonlinear.Limiter.fmu", "limiter")
session.instantiateFMU(model, "../FMUs/Modelica.Blocks.Math.Add.fmu", "addSat")
session.instantiateFMU(model, "../FMUs/Modelica.Blocks.Math.Gain.fmu", "gainTrack")

# add connections
session.addConnection(model, "addP.y", "P.u")
session.addConnection(model, "addI.y", "I.u")
session.addConnection(model, "P.y", "addPI.u1")
session.addConnection(model, "I.y", "addPI.u2")
session.addConnection(model, "addPI.y", "gainPI.u")
session.addConnection(model, "gainPI.y", "limiter.u")
session.addConnection(model, "gainPI.y", "addSat.u2")
session.addConnection(model, "limiter.y", "addSat.u1")
session.addConnection(model, "addSat.y", "gainTrack.u")
session.addConnection(model, "gainTrack.y", "addI.u3")

# parameters
k = 100.0
yMax = 12.0
yMin = -yMax
wp = 1.0
Ni = 0.1
xi_start = 0.0

# set parameters
session.setReal(model, "addP.k1", wp)
session.setReal(model, "addP.k2", -1.0)
session.setReal(model, "addI.k2", -1.0)
session.setReal(model, "I.y_start", xi_start)
session.setReal(model, "gainPI.k", k)
session.setReal(model, "limiter.uMax", yMax)
session.setReal(model, "addSat.k2", -1.0)
session.setReal(model, "gainTrack.k", 1.0/(k*Ni))

# simulation settings
session.setStopTime(model, 4.0)
session.setCommunicationInterval(model, 1e-2)

session.exportCompositeStructure(model, "PI_Controller")
session.exportDependencyGraph(model, "PI_Controller")
#gvpr -c "N[$.degree==0]{delete(root, $)}" PI_Controller_simulation.dot | dot -Tpdf -o PI_Controller_simulation.pdf
#gvpr -c "N[$.degree==0]{delete(root, $)}" PI_Controller_initialization.dot | dot -Tpdf -o PI_Controller_initialization.pdf
#dot -Gsplines=none PI_Controller.dot | neato -n -Gsplines=ortho -Tpdf -oPI_Controller.pdf

# instantiate lookup table
session.instantiateTable(model, "setpoint.csv", "setpoint")
session.instantiateTable(model, "drivetrain.csv", "driveTrain")

# add connections
session.addConnection(model, "setpoint.speed", "addP.u1")
session.addConnection(model, "setpoint.speed", "addI.u1")

session.addConnection(model, "driveTrain.w", "addP.u2")
session.addConnection(model, "driveTrain.w", "addI.u2")

# simulation settings
session.setResultFile(model, "PI_Controller.mat")

# simulation
print("Initialization")
session.initialize(model)
print("  limiter.u: " + str(session.getReal(model, "limiter.u")))
print("  limiter.y: " + str(session.getReal(model, "limiter.y")))

print("Simulation")
session.simulate(model)
print("  limiter.u: " + str(session.getReal(model, "limiter.u")))
print("  limiter.y: " + str(session.getReal(model, "limiter.y")))

session.terminate(model)
session.unload(model)

vars = ["limiter.u", "limiter.y"]
for var in vars:
  if (1 == session.compareSimulationResults("PI_Controller.mat", "../ReferenceFiles/PI_Controller.mat", var, 1e-2, 1e-4)):
    print(var + " is equal")
  else:
    print(var + " is not equal")

## Result:
## Initialization
##   limiter.u: 0.0
##   limiter.y: 0.0
## Simulation
##   limiter.u: -10.3343691227
##   limiter.y: -10.3343691227
## limiter.u is not equal
## limiter.y is not equal
## info:    4 warnings
## info:    0 errors
## info:    Logging information has been saved to "PI_Controller.log"
## endResult
